### ISSUE TITLE
GCP pub/sub MQL no longer supported(recommend PromQL) - Starting October 22, 2024

### ISSUE THREAD

**Issue by yazimpact** [kedacore/keda#6255]
### Report

GCP Monitoring Page saw below warning: 

> Starting October 22, 2024, Monitoring Query Language (MQL) will no longer be a recommended query language for Cloud Monitoring, and we will begin to turn off certain usability features. We recommend moving to [PromQL ](https://cloud.google.com/monitoring/promql), the open-source standard for querying time series. PromQL offers similar functionality to MQL, with a wider user base and more community resources. Learn more

### Expected Behavior

Use `Promql` instead of `MQL`

### Actual Behavior

Current metrics is using MQL:

```
pubsub.googleapis.com/subscription/num_undelivered_messages
| filter (resource.project_id == 'project-name' && resource.subscription_id == 'example-sub')
| within 1m
```

### Steps to Reproduce the Problem

References: 
**PROMQL**:
https://cloud.google.com/monitoring/promql
**MQL**:
https://cloud.google.com/stackdriver/docs/deprecations/mql

### Logs from KEDA operator

```
example
```

### KEDA Version

2.13.1

### Kubernetes Version

1.30

### Platform

Google Cloud

### Scaler Details

GCP Pub/Sub

### Anything else?

https://cloud.google.com/stackdriver/docs/deprecations/mql 

On October 22, 2024, [Monitoring Query Language (MQL)](https://cloud.google.com/monitoring/mql) will no longer be a recommended query language for Cloud Monitoring.

On October 22, 2024, certain usability features will be turned off.
On July 22, 2025, MQL will no longer be available for new dashboards and alerts in the Google Cloud console, and Google Cloud customer support will end. Existing MQL dashboards and alerts will continue to work, and you will still be able to create MQL dashboards and alerts using the Cloud Monitoring API.

**Comment by JorTurFer** [comment]
Interesting update :smile:

@kedacore/keda-core-contributors , do you think that we should get rid of the GCP Scalers or just use them as wrapper on top of the prometheus one to make the things easier to end users?

**Comment by yazimpact** [comment]
Not exactly sure what could be the best approach for users.

**Comment by yazimpact** [comment]
@JorTurFer  any updated? Keda started getting errors for scaling:

```
github.com/kedacore/keda/v2/pkg/scalers.(*pubsubScaler).GetMetricsAndActivity
        /workspace/pkg/scalers/gcp_pubsub_scaler.go:193
github.com/kedacore/keda/v2/pkg/scaling/cache.(*ScalersCache).GetMetricsAndActivityForScaler
        /workspace/pkg/scaling/cache/scalers_cache.go:140
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScalerState
        /workspace/pkg/scaling/scale_handler.go:743
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScaledObjectState.func1
        /workspace/pkg/scaling/scale_handler.go:628
2024-10-23T11:08:11Z    ERROR   scale_handler   error getting scale decision    {"scaledObject.Namespace": "pap-2", "scaledObject.Name": "example-scaledobject", "scaler": "pubsubScaler", "error": "could not find stackdriver metric with query fetch pubsub_subscription | metric 'pubsub.googleapis.com/subscription/num_undelivered_messages' | filter (resource.project_id == 'project_name' && resource.subscription_id == 'example-sub') | within 1m"}
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScalerState
        /workspace/pkg/scaling/scale_handler.go:764
github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).getScaledObjectState.func1
        /workspace/pkg/scaling/scale_handler.go:628
```

**Comment by wozniakjan** [comment]
am I reading https://cloud.google.com/stackdriver/docs/deprecations/mql correctly by expecting MQL to work until July 22, 2025?

also the same document mentions:
> No action is required to continue using your existing MQL assets, and you can create new MQL assets by using the Cloud Monitoring API.

**Comment by JorTurFer** [comment]
I'm not 100% sure about the right solution, let's talk about this during next standup @zroubalik @wozniakjan

**Comment by JorTurFer** [comment]
I agree with that, IMHO we should deprecate the GCP scalers and document how to use the prometheus scaler as replacement

**Comment by yazimpact** [comment]
I have not tested yet with GCP Managed Prometheus as scaler, but will be doing in near future just in case(if mql completely gone).

**Comment by sunhubs** [comment]
> I agree with that, IMHO we should deprecate the GCP scalers and document how to use the prometheus scaler as replacement

where is the guide?

**Comment by rickbrouwer** [comment]
July 25th is fast approaching, and so is the release of v2.18. What are the plans or consensus about the scalers that use MQL and the possible deprecation of these scalers? This because the deprecation message might still be included in the v2.18 release if we want.

**Comment by JorTurFer** [comment]
I'm reading the things with calm, and it looks that the API will be the same, the stackdriver one, but just the querying language changes, so maybe pushing users to prometheus is overkill, maybe we should explore the option of using the current driver just with a PromQL like query

**Comment by rickbrouwer** [comment]
Yes, that seems like a good plan to me.
When I look at the page, it says the following:

> No action is required to continue using your existing MQL assets, and you can create new MQL assets by using the Cloud Monitoring API.

So, essentially, "no action is required," but it's recommended to use PromQL.

> We recommend that you begin using PromQL [...]

Knowing this, I think we might need to support both, as there's no roadmap yet (at least, I don't see one) for truly phasing out MQL.

**Comment by JorTurFer** [comment]
We don't expose the underlaying query to endusers, they just introduce the metrics and conditions, and KEDA builds de query using the querying language. That's why I said that maybe we can  switch to PromQL the underlaying query. Maybe we can include a temporal flag to use PromQL querying model and call for feedback during 1-2 versions (and it also gives us a workaround in case of not announced support removal)

**Comment by JorTurFer** [comment]
I've been testing the change in [this PR](https://github.com/kedacore/keda/pull/6968/files) and the change isn't just update the querying. [The API that has to be used changes](https://discuss.google.dev/t/promql-monitoring-queries-via-api/109343), so using the Prometheus Scaler is the best way. Without enabling any extra services, I've been able to query the Cloud Monitoring Prom API (no extra Prometheus is needed)
To keep the current scalers based on stackdriver, we would need to adapt them and call to Prometheus endpoints using Prometheus client. 
Considering this, I'm going to open another PR deprecating the current GCP Scalers affected by this and another PR explaining how to migrate from the current scalers to the Prometheus one. I'd not remove the current scaler [because google says](https://cloud.google.com/stackdriver/docs/deprecations/mql) that *MQL is not being shut down*, but seeing how google closes projects, I think that deprecating the current MQL scalers and suggest users to move to PromQL is worth for us.
WDYT @kedacore/keda-core-contributors ?

**Comment by zroubalik** [comment]
Agree

**Comment by michaelbannister** [comment]
I totally understand that from the maintainers' point of view, this looks like a great opportunity to simplify, reuse the existing Prometheus scaler. But from a user's point of view, this means they have to know more about the specifics, how to form the PromQL correctly, etc. Even if familiar with PromQL, Stackdriver's use of gauges and deltas rather than Prometheus-style counters means the translation can be confusing.

**Comment by JorTurFer** [comment]
Hello
That's a real good point! We won't remove the the current GCP scalers so current users aren't forced to migrate if they don't want. The point  is that sometimes there isn't direct translation from MQL to PromQL, so se can't just hide all the magic under the hood.
Our idea is to keep the current GCP scalers based on MQL meanwhile GCP supports it, we just won't include more features, but they will continue as they are.

More info here: https://keda.sh/blog/2025-09-15-gcp-deprecations/
